crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +472,29 @@ impl<'db> Type<'db> {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct IterableElementTypeResult<'db> {
+    element_ty: Option<Type<'db>>,
+    iterable_ty: Type<'db>,
+}
```

Comment:
I would use an `enum` here because `iterable_ty` is misleading when it is actually a non-iterable ty.

```rust
enum IterateType<'db> {
	Iterable { element_ty: Type<'db> },
	NonIterable { ty: Type<'db> }
}
```

It also makes it more obvious that only one type is used at a time.